lib/Sema/CSApply.cpp

```diff
@@ -2862,11 +2862,67 @@ namespace {
       // Check for ambiguous member if the base is an Optional
       if (baseTy->getOptionalObjectType()) {
         diagnoseAmbiguousNominalMember(baseTy, result);
+      } else {
+        tryDiagnoseContextualOptionalNoneMemberAmbiguous(expr, baseTy);
       }
 
       return coerceToType(result, resultTy, cs.getConstraintLocator(expr));
     }
-    
+
+    /// Diagnose an unresolved `.none` member infered from a contextual optional
+    /// type where the member base type is infered as a static non-optional member.
+    /// So we enforce explicit type anotation to avoid ambiguity.
+    ///   enum Enum<T> {
+    ///     case bar
+    ///     static var none: Enum<Int> { .bar }
+    ///   }
+    ///   let _: Enum? = .none // Infered as static member Enum<Int>.none, emit
+    ///   warning suggesting explicit type.
+    ///   let _: Enum = .none // Ok
+    ///
+    void tryDiagnoseContextualOptionalNoneMemberAmbiguous(UnresolvedMemberExpr *UME,
+                                                          Type baseTy) {
+      assert(!baseTy->getOptionalObjectType());
+      
+      if (!UME->getName().isSimpleName("none"))
+        return;
+      
+      // Return if the base type doesn't have a nominal type decl
+      if (!baseTy->getNominalOrBoundGenericNominal())
+        return;
+      
+      // Find member chain result in order to get contextual type.
+      auto *parent = dyn_cast_or_null<UnresolvedMemberChainResultExpr>(
+          cs.getParentExpr(UME));
+      if (!parent)
+        return;
+
+      if (auto contextualType = cs.getContextualType(parent)) {
```

Comment:
I think this should be `cs.getType(parent)` since we might not have an actual contextual type guiding inference (e.g., when `.none` is in function argument position).

Reply:
`cs.getType(parent)` is the inferred type for type variable representing the member type, and I think that is not what we want here. We want to know if for this type variable an optional type `Optional.none` binding was attempted (meaning it could be inferred as Optional) ... which is the information we get from the contextual type. And I'm actually not sure we have a way to get this information right now (one way may be tracking the source of the type variable binding)

Reply:
IIRC, `cs.getType(UME)` would be the type of the member access, but `cs.getType(parent)` (after simplification) should give us the type that the member type will ultimately be converted to. E.g. with:
```
struct Foo<T> {
    static var none: Foo<Int> { Foo<Int>() }
}

let _: Foo? = .none
```

The `UnresolvedMemberChainResultExpr` ends up with type `Foo<Int>?`.

Reply:
Seems like that is not the case
```
(lldb) p parentTy->dump()
(bound_generic_struct_type decl=main.(file).Foo@path:3:8
  (struct_type decl=Swift.(file).Int))
(lldb) p contextualType->dump()
(optional_type
  (unbound_generic_type decl=main.(file).Foo@path:3:8))
(lldb) 
```
Being `parentType = simplifyType(cs.getType(parent))`, but I'm taking a deeper look.

Also thank you for joining together on this :)

Reply:
Ah, I realize what's happening! I was testing against my branch for #34715, where we attempt both overloads on the binding step where we bind `Foo<_>?` to the `UnresolvedMemberChainResultExpr`. Without those changes, we only attempt the `Foo.none` overload when the `Foo<_>?` binding step fails and we attempt `Foo<_>`.

So you are correct that without those changes, the solution here will bind the `UnresolvedMemberChainResultExpr` to non-optional `Foo`. :)

Reply:
Ah I see, thanks 
I just change this to make this check based on conversion restrictions and diagnose only if an inferred type has a value to optional conversion restriction e.g. `Foo<Int> to Optional<Foo<Int>>` :)